### ISSUE TITLE
Package Update Request: sys-libs/glibc 2.32

### ISSUE THREAD

**Issue by dongsupark** [flatcar/Flatcar#145]
**sys-libs/glibc 2.32-r8**

Recently GLSA published multiple security vulnerabilities [GLSA 202006-04](https://security.gentoo.org/glsa/202006-04) in `sys-libs/glibc`.
For example, it is about [CVE-2019-6488](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-6488).

It is recommended to update glibc to 2.32 or newer. At the moment Flatcar has glibc 2.29, which is obviously affected.

By the way, there is also a well-known issue in Flatcar Alpha/Edge. `virtual/libcrypt` 2 or newer started to depend on `sys-libs/libxcrypt`, which does not exist in Flatcar. However, `libxcrypt` depends on `glibc` with the USE flag `crypt`, which is only available in glibc 2.30 or newer. So it has not been possible to add `sys-libs/libxcrypt`, and `virtual/libcrypt` gets still installed as version 1, printing out noisy warning messages. This is also one of the reasons why we need glibc 2.32.

However, on the other hand, `sys-libs/glibc` ebuilds have some third-party patches created by ex-coreos maintainers. So we need to port the patches to make it work with glibc 2.32.

- The package improves on the following core values:
  - [x] Secure by default
  - [x] Always up to date

Update 2020-09-30: update to 2.32 seems feasible, since gcc needs to be updated to 9.3 anyways. Changed title and description accordingly.

**Comment by dongsupark** [comment]
Filed a PR for alpha. https://github.com/flatcar-linux/coreos-overlay/pull/431

**Comment by t-lo** [comment]
Thank you for the great work leading here, Dongsu!
New WIP branches w/ toolchain updates:
- https://github.com/flatcar-linux/coreos-overlay/tree/t-lo/update-glibc-gcc-tools (glibc etc.)
- https://github.com/flatcar-linux/portage-stable/tree/t-lo/update-glibc-gcc-tools (gcc, binutils, etc.)

**Comment by t-lo** [comment]
Updated nss-altfiles with a [patch](https://github.com/flatcar-linux/nss-altfiles/commit/9078c543ba7d2bc5011737675b3dddb882673ce7) to un-break for glibc-2.30 or newer.

**Comment by t-lo** [comment]
The SDK build breaks when trying to build the old gcc, 8.3.0, with a libsanitizer error:
```
In file included from [01m[K/var/tmp/portage/sys-devel/gcc-8.3.0/work/gcc-8.3.0/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc:193[m[K:
[01m[K/var/tmp/portage/sys-devel/gcc-8.3.0/work/gcc-8.3.0/libsanitizer/sanitizer_common/sanitizer_internal_defs.h:317:72:[m[K [01;31m[Kerror: [m[Ksize of array '[01m[Kassertion_failed__1152[m[K' is negative
     typedef char IMPL_PASTE(assertion_failed_##_, line)[2*(int)(pred)-1[01;31m[K][m[K
                                                                        [01;31m[K^[m[K
[01m[K/var/tmp/portage/sys-devel/gcc-8.3.0/work/gcc-8.3.0/libsanitizer/sanitizer_common/sanitizer_internal_defs.h:311:30:[m[K [01;36m[Knote: [m[Kin expansion of macro '[01m[KIMPL_COMPILER_ASSERT[m[K'
 #define COMPILER_CHECK(pred) [01;36m[KIMPL_COMPILER_ASSERT[m[K(pred, __LINE__)
                              [01;36m[K^~~~~~~~~~~~~~~~~~~~[m[K
[01m[K/var/tmp/portage/sys-devel/gcc-8.3.0/work/gcc-8.3.0/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.h:1475:3:[m[K [01;36m[Knote: [m[Kin expansion of macro '[01m[KCOMPILER_CHECK[m[K'
   [01;36m[KCOMPILER_CHECK[m[K(sizeof(((__sanitizer_##CLASS *) NULL)->MEMBER) == \
   [01;36m[K^~~~~~~~~~~~~~[m[K
[01m[K/var/tmp/portage/sys-devel/gcc-8.3.0/work/gcc-8.3.0/libsanitizer/sanitizer_common/sanitizer_platform_limits_posix.cc:1152:1:[m[K [01;36m[Knote: [m[Kin expansion of macro '[01m[KCHECK_SIZE_AND_OFFSET[m[K'
 [01;36m[KCHECK_SIZE_AND_OFFSET[m[K(ipc_perm, mode);
 [01;36m[K^~~~~~~~~~~~~~~~~~~~~
```

**Comment by t-lo** [comment]
We'll first update gcc & friends (crossdev, binutils, etc.) rebuild the SDK, then update glibc.

**Comment by t-lo** [comment]
The old gcc's libsanitizer fails because of an array of negative size; details in https://github.com/flatcar-linux/Flatcar/issues/145 .

glibc-2.30 and 2.32 have been demonstrated to compile when installing gcc-9.3.0 manually on top of an existing SDK (i.e. updating `cross-x86_64-cros-linux-gnu/gcc` from 8.3.0 to 9.3.9 and building an OS image w/ glibc-2.30).

**Comment by t-lo** [comment]
Updated perf to make it work with new binutils.
Hitting https://bugs.gentoo.org/690788 after updating kernel-headers. Will update dnsmasq to 2.28 to mitigate.

**Comment by t-lo** [comment]
SDK build succeeded (gcc-9.3.0 and binutils-2.35) but toolchains build now fails with a circular dependency similar to https://bugs.gentoo.org/630880 .

**Comment by t-lo** [comment]
Backporting gcc and binutils recipes to EAPI 6 resolved the circular dep.

Strace needs to be updated to 5.8 or newer since the build breaks against the new kernel headers (reference e.g. here https://patchwork.openembedded.org/patch/175096/).

**Comment by t-lo** [comment]
SDK, image and toolchains build successfully with gcc-9.3.0 and binutils-2.35. Now that we have a working gcc-9.3 SDK we can ingest glibc-2.32.

**Comment by t-lo** [comment]
PRs for updating the compiler to gcc-9.3.0 (https://github.com/kinvolk/portage-stable/pull/127, https://github.com/kinvolk/coreos-overlay/pull/693) tested successfully, merged to main.

**Comment by t-lo** [comment]
Alpha-2697.0.0 SDK ships gcc-9.3.0 and binutils-2.35. Will now ingest glibc-2.32.

**Comment by t-lo** [comment]
WIP branches are https://github.com/kinvolk/coreos-overlay/tree/t-lo/update-to-glibc-2.32 and https://github.com/kinvolk/portage-stable/tree/t-lo/update-glibc-to-2.32 .

**Comment by t-lo** [comment]
We need to update samba to 4.12 because we're hitting https://forums.gentoo.org/viewtopic-t-1118902-start-0.html with glibc-2.32.

**Comment by t-lo** [comment]
PRs merged to main; glibc-2.32 will be included in the upcoming Alpha release.

**Comment by t-lo** [comment]
Included in [Alpha-2705.0.0](https://www.flatcar-linux.org/releases/#release-2705.0.0).
